### PR TITLE
feat(provider)!: remove support for python2 and python3.[3-6]

### PR DESCRIPTION
**Pull request by bfredl**
These versions of python has reached End-of-life. getting rid of python2 support removes a lot of logic to support two
incompatible python versions in the same version.

Note: this is a draft but not marked as such because I want to see all mayhem this causes in CI at once.

Note: as currently `has("python")` will return false while `has("python3")` and `has("pythonx")` will return true when available. I am inclined to _not_ change this as this is the _only_ way plugins can check that python2 is not supported without new special logic for neovim 0.7+ only. However if the user/plugin uses `:python` / `pyeval()` without checking `has()`  we **will** give them python, that is python3, as they asked, as python now unambiguously means python3.

### CODE REVIEW

**Review by clason — APPROVED** [review]

**Comment by fsouza** [comment]
3.6 is technically EOL too, but I guess that's too soon?

File: runtime/doc/if_pyth.txt [hunk 1]
```diff
@@ -590,6 +589,11 @@ functions to evaluate Python expressions and pass their values to Vim script.
 ==============================================================================
 Python 3						*python3*
 
+As python3 is the only supported version in Nvim, "python" is synonymous
```
Comment (clason) [inline runtime/doc/if_pyth.txt:592]: Consistent phrasing in this paragraph: "Python 3", "Nvim"

File: runtime/doc/if_pyth.txt [hunk 2]
```diff
@@ -619,66 +623,43 @@ Raising SystemExit exception in python isn't endorsed way to quit vim, use: >
...
-	  echo 'there is Python 2.x'
+You can test if Python is available with: >
+	if has('pythonx')
+	  echo 'there is Python'
 	endif
   	if has('python3')
 	  echo 'there is Python 3.x'
 	endif
 
+Python 2 is no longer supported. thus `has('python')` always returns
```
Comment (clason) [inline runtime/doc/if_pyth.txt:634]: "Thus" (capitalize).

File: runtime/doc/if_pyth.txt [hunk 3]
```diff
@@ -619,66 +623,43 @@ Raising SystemExit exception in python isn't endorsed way to quit vim, use: >
...
-'pyxversion' setting.
-
-Set 'pyxversion' in your |vimrc| to prefer Python 2 or Python 3 for Python
-commands. Changing this setting at runtime risks losing the state of plugins
-(such as initialization).
-
-If you want to use a module, you can put it in the {rtp}/pythonx directory.
-See |pythonx-directory|.
+The "pythonx" and "pyx" prefixes was introduced for python code which
+works with Python 2.6+ and Python 3. As Nvim only supports python3,
```
Comment (clason) [inline runtime/doc/if_pyth.txt:641]: "Python 3

File: runtime/doc/if_pyth.txt [hunk 4]
```diff
@@ -590,6 +589,11 @@ functions to evaluate Python expressions and pass their values to Vim script.
 ==============================================================================
 Python 3						*python3*
 
+As python3 is the only supported version in Nvim, "python" is synonymous
+with "python3" in the current version. However, code that aims to support older
+versions of Neovim, as well as Vim, should prefer to use use "python3"
```
Comment (mattpolzin) [inline runtime/doc/if_pyth.txt:594]:
```suggestion
versions of Neovim, as well as Vim, should prefer to use "python3"
```
extra word spotting 🙂

**Comment by bfredl** [comment]
@fsouza perhaps we should. I was just going by https://github.com/neovim/pynvim/pull/492 at this point :]

**Comment by bfredl** [comment]
Tests are updated and doc reviews addressed.

File: runtime/doc/if_pyth.txt [hunk 5]
```diff
@@ -619,66 +623,43 @@ Raising SystemExit exception in python isn't endorsed way to quit vim, use: >
...
+You can test if Python is available with: >
+	if has('pythonx')
+	  echo 'there is Python'
 	endif
   	if has('python3')
 	  echo 'there is Python 3.x'
 	endif
 
+Python 2 is no longer supported. Thus `has('python')` always returns
+zero for backwards compatibilityTreasons.
```
Comment (clason) [inline runtime/doc/if_pyth.txt:635]:
```suggestion
zero for backwards compatibility reasons.
```

File: runtime/doc/if_pyth.txt [hunk 6]
```diff
@@ -619,66 +623,43 @@ Raising SystemExit exception in python isn't endorsed way to quit vim, use: >
...
-
-Set 'pyxversion' in your |vimrc| to prefer Python 2 or Python 3 for Python
-commands. Changing this setting at runtime risks losing the state of plugins
-(such as initialization).
-
-If you want to use a module, you can put it in the {rtp}/pythonx directory.
-See |pythonx-directory|.
+The "pythonx" and "pyx" prefixes were introduced for python code which
+works with Python 2.6+ and Python 3. As Nvim only supports Python 3,
+all these commands are now synonymous to their "python3" equivalents
```
Comment (clason) [inline runtime/doc/if_pyth.txt:642]:
```suggestion
all these commands are now synonymous to their "python3" equivalents.
```

File: runtime/doc/if_pyth.txt [hunk 7]
```diff
@@ -619,66 +623,43 @@ Raising SystemExit exception in python isn't endorsed way to quit vim, use: >
...
-If Vim does not support the selected Python version a silent message will be
-printed.  Use `:messages` to read them.
+`:pyxfile` works the same as `:py3file`. 
 
 							*:pyxdo*
-`:pyxdo` works similar to `:pydo`.
+`:pyxdo` works the same as `:py3do`.
 
 							*has-pythonx*
 To check if pyx* functions and commands are available: >
```
Comment (clason) [inline runtime/doc/if_pyth.txt:659]:
```suggestion
To check if `:pyx*` functions and commands are available: >
```

File: runtime/doc/options.txt [hunk 8]
```diff
@@ -4637,26 +4637,11 @@ A jump table for the options with a short description can be found at |Q_op|.
...
-	If 'pyxversion' is 0, it is set to 2 or 3 after the first execution of
-	any python2/3 commands or functions.  E.g. `:py` sets to 2, and `:py3`
-	sets to 3. `:pyx` sets it to 3 if Python 3 is available, otherwise sets
-	to 2 if Python 2 is available.
-	See also: |has-pythonx|
-
-	If only |+python| or |+python3| are available,
-	'pyxversion' has no effect. The pyx* functions and commands are
-	always the same as the installed version.
+	|python_x|.  As only python3 is supported in this version,
```
Comment (clason) [inline runtime/doc/options.txt:4643]:
```suggestion
	|python_x|.  As only Python 3 is supported,
```

File: runtime/doc/provider.txt [hunk 9]
```diff
@@ -20,11 +20,12 @@ Run the |:checkhealth| command, and review the sections below.
...
 Python integration				*provider-python*
 
-Nvim supports Python |remote-plugin|s and the Vim legacy |python2| and
-|python3| interfaces (which are implemented as remote-plugins).
+Nvim supports Python |remote-plugin|s and the Vim legacy |python3| and
+|pythonx| interfaces (which are implemented as remote-plugins).
 
 Note: Only the Vim 7.3 legacy interface is supported, not later features such
-as |python-bindeval| (Vim 7.4); use the Nvim API instead.
+as |python-bindeval| (Vim 7.4); use the Nvim API instead. Python2 is not
```
Comment (clason) [inline runtime/doc/provider.txt:27]:
```suggestion
as |python-bindeval| (Vim 7.4); use the Nvim API instead. Python 2 is not
```

**Review by clason — APPROVED** [review]
Minor docnits, otherwise LGTM. (Love those diffstats :])

File: runtime/doc/if_pyth.txt (hunk 7 above)
Comment (bfredl) [reply runtime/doc/if_pyth.txt:659]: they are functions as well tho :]

**Comment by mckellygit** [comment]
With this commit I now see this on every startup -
```
Error detected while processing function provider#pythonx#Detect[1]..provider#pythonx#DetectByModule[1]..<SNR>28_get_python_executable_from_host_var:
line    1:
E605: Exception not caught: unsupported
Press ENTER or type command to continue
```
But I think my python is ok ?
```
## Python 3 provider (optional)
  - INFO: Using: g:python3_host_prog = "/usr/bin/python3.8"
  - INFO: Executable: /usr/bin/python3.8
  - INFO: Python version: 3.8.10
  - INFO: pynvim version: 0.4.3
  - OK: Latest pynvim is installed.

## Python virtualenv
  - OK: no $VIRTUAL_ENV
```
and
```
% python --version
Python 3.8.10
% python3 --version
Python 3.8.10
```
This seems to be from pythonx.vim -
```
function! s:get_python_executable_from_host_var(major_version) abort
  return expand(get(g:, 'python'.(a:major_version == 3 ? '3' : execute("throw 'unsupported'")).'_host_prog', ''), v:true)
endfunction
```
Do you know where the major_version could be getting set (to 2) ?
Is it a plugin I have that might be looking for python2 ?
Do I need to remove all python2 from my system ?
thx,
-m

**Comment by mckellygit** [comment]
ok, I think its the vimade plugin.  I will look into that and work with that developer.
thx again for nvim,
-m

**Comment by yaegassy** [comment]
I am using "vim-plug" as my plugin manager. Since this PR was merged, I can't use `:PlugInstall` or `:PlugUpdate` with "vim-plug".

I found out that it is because `has('python')` raises an error the first time it is executed.

There may be a situation where all plugins that use `has('python')` are broken. :(

**Comment by bfredl** [comment]
You might need to clean out stale runtime files like this suggests. https://github.com/neovim/neovim/pull/17252#issuecomment-1025488419

**Comment by yaegassy** [comment]
I always build Neovim with `CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/neovim" CMAKE_BUILD_TYPE=Release`. I deleted `$HOME/neovim` and built it, and the problem was solved.

**Comment by mckellygit** [comment]
Thanks so much.
I also just now did a rm -rf /usr/local/share/nvim/runtime before a make install and no longer see the issue.
-m

**Pull request merged**